protocols/relay/src/v2/relay.rs

```diff
@@ -0,0 +1,678 @@
+    },
+    /// Accepting an inbound circuit request failed.
+    CircuitReqAcceptFailed {
+        src_peer_id: PeerId,
+        dst_peer_id: PeerId,
+        error: std::io::Error,
+    },
+    /// An inbound circuit has closed.
+    CircuitClosed {
+        src_peer_id: PeerId,
+        dst_peer_id: PeerId,
+        error: Option<std::io::Error>,
+    },
+}
+
+/// [`Relay`] is a [`NetworkBehaviour`] that implements the relay server
+/// functionality of the circuit relay v2 protocol.
+pub struct Relay {
+    config: Config,
+
+    local_peer_id: PeerId,
+
+    reservations: HashMap<PeerId, HashSet<ConnectionId>>,
+    circuits: CircuitsTracker,
+
+    /// Queue of actions to return when polled.
+    queued_actions: VecDeque<Action>,
+}
+
+impl Relay {
+    pub fn new(local_peer_id: PeerId, config: Config) -> Self {
+        Self {
+            config,
+            local_peer_id,
+            reservations: Default::default(),
+            circuits: Default::default(),
+            queued_actions: Default::default(),
+        }
+    }
+}
+
+impl NetworkBehaviour for Relay {
+    type ProtocolsHandler = handler::Prototype;
+    type OutEvent = Event;
+
+    fn new_handler(&mut self) -> Self::ProtocolsHandler {
+        handler::Prototype {
+            config: handler::Config {
+                reservation_duration: self.config.reservation_duration,
+                max_circuit_duration: self.config.max_circuit_duration,
+                max_circuit_bytes: self.config.max_circuit_bytes,
+            },
+        }
+    }
+
+    fn inject_connection_closed(
+        &mut self,
+        peer: &PeerId,
+        connection: &ConnectionId,
+        _: &ConnectedPoint,
+        _handler: handler::Handler,
+    ) {
+        self.reservations
+            .get_mut(peer)
+            .map(|cs| cs.remove(&connection))
+            .unwrap_or(false);
+
+        for circuit in self
+            .circuits
+            .remove_by_connection(*peer, *connection)
+            .iter()
+            // Only emit [`CircuitClosed`] for accepted requests.
+            .filter(|c| matches!(c.status, CircuitStatus::Accepted))
+        {
+            self.queued_actions.push_back(
+                NetworkBehaviourAction::GenerateEvent(Event::CircuitClosed {
+                    src_peer_id: circuit.src_peer_id,
+                    dst_peer_id: circuit.dst_peer_id,
+                    error: Some(std::io::ErrorKind::ConnectionAborted.into()),
+                })
+                .into(),
+            );
+        }
+    }
+
+    fn inject_event(
+        &mut self,
+        event_source: PeerId,
+        connection: ConnectionId,
+        event: handler::Event,
+    ) {
+        match event {
+            handler::Event::ReservationReqReceived {
+                inbound_reservation_req,
+                remote_addr,
+            } => {
+                let now = Instant::now();
+
+                let action = if remote_addr.iter().any(|p| p == Protocol::P2pCircuit) {
```

Comment:
If this was already a relayed connection, we would be the destination node (= a client). If I understand the code correctly, we would not observed the dialer/ source peer on a multi-address that includes `Protocol::P2pCircuit`, but instead our client transport would just set `remote_addr`  to just a single `Protocol::P2p(src_id)`? 
https://github.com/mxinden/rust-libp2p/blob/dbfe25bf5872700a281d058e62b4adf822f10d56/protocols/relay/src/v2/client/transport.rs#L312
Am I missing something here? Same in line 341.

Reply:
:fearful: great catch! Fixed via https://github.com/libp2p/rust-libp2p/pull/2059/commits/451d86c36b0cbcca637bf0b1578e891d2542c313. Potentially simplified via https://github.com/libp2p/rust-libp2p/pull/2392.

Let me know what you think of the solution @elenaf9.

Reply:
Looks good to me!